### PR TITLE
Add more metrics around finalization attempts

### PR DESCRIPTION
**Pull request by kelvinfan001**
This adds metrics to measure the total number of finalization
attempts, `can_finalize` responses from  update strategy, and
attempts denied due to factors unrelated to update strategy,
such as active user sessions, or (in the future) systemd inhibitor
locks present.

Note: this is in addition to and distinct from the
`zincati_rpm_ostree_finalize_attempts_total`
`zincati_rpm_ostree_finalize_failures_total` metrics, as those two
metrics are solely related to the `rpm_ostree` actor and rpm-ostree
binary.

Closes https://github.com/coreos/zincati/issues/544.

### CODE REVIEW

File: src/update_agent/actor.rs [hunk 1]
```diff
@@ -16,6 +16,26 @@ lazy_static::lazy_static! {
         "zincati_update_agent_last_refresh_timestamp",
         "UTC timestamp of update-agent last refresh tick."
     )).unwrap();
+    static ref FINALIZATION_ATTEMPTS: IntCounter = register_int_counter!(opts!(
+        "zincati_update_agent_finalization_attempts",
+        "Total number of attempts to finalize a staged deployment by the update agent."
+    )).unwrap();
+    static ref FINALIZATION_ALLOWED_BY_STRATEGY: IntCounter = register_int_counter!(opts!(
```
Comment (lucab) [inline src/update_agent/actor.rs:23]: I think this allowed/denied can be folded into a single metrics with a `outcome=allowed|denied` label and moved into the `strategy` module. I'm not sure whether to fold the error outcome into this label as well.

File: src/update_agent/actor.rs [hunk 2]
```diff
@@ -16,6 +16,26 @@ lazy_static::lazy_static! {
...
+        "zincati_update_agent_finalization_allowed_by_strategy",
+        "Total number of finalization attempts by the update agent allowed by its update strategy."
+    )).unwrap();
+    static ref FINALIZATION_DENIED_BY_STRATEGY: IntCounter = register_int_counter!(opts!(
```
Comment (lucab) [inline src/update_agent/actor.rs:27]: These `denied` cases can also likely be folded into labels.

How do you plan to fit inhibitors into the flow (and thus metrics counting)?

File: src/update_agent/actor.rs (hunk 1 above)
Comment (kelvinfan001) [reply src/update_agent/actor.rs:23]: By "folding", you mean using e.g. `IntCounterVec` right?

File: src/update_agent/actor.rs (hunk 2 above)
Comment (kelvinfan001) [reply src/update_agent/actor.rs:27]: I was thinking of making a separate metric, similar to `zincati_update_agent_finalization_denied_by_usersessions` to count denials due to systemd inhibitors. Should we combine all non strategy-related gates into one?

File: src/update_agent/actor.rs (hunk 1 above)
Comment (lucab) [reply src/update_agent/actor.rs:23]: Yes. There are some examples in this project in the `fleet_lock` module.

File: src/update_agent/actor.rs (hunk 2 above)
Comment (lucab) [reply src/update_agent/actor.rs:27]: Yes, I'd combine them together via labels.

**Comment by kelvinfan001** [comment]
@lucab I think this is good for another look if you have time :)

**Review by lucab — APPROVED** [review]

**Pull request merged**